src/blockencodings.cpp

```diff
@@ -0,0 +1,170 @@
+                // This should be rare enough that the extra bandwidth doesn't matter,
+                // but eating a round-trip due to FillBlock failure would be annoying
+                pool->ReleaseTxLock(txhashes[idit->second]);
+                txhashes[idit->second].SetNull();
+            }
+        }
+    }
+
+    LogPrint("cmpctblock", "Initialized PartiallyDownloadedBlock for block %s using a cmpctblock of size %lu\n", cmpctblock.header.GetHash().ToString(), cmpctblock.GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION));
+
+    return READ_STATUS_OK;
+}
+
+bool PartiallyDownloadedBlock::IsTxAvailable(size_t index) const {
+    assert(!header.IsNull());
+    assert(index < txhashes.size());
+    return !txhashes[index].IsNull();
+}
+
+ReadStatus PartiallyDownloadedBlock::FillBlock(CBlock& block, const std::vector<CTransaction>& vtx_missing) const {
+    assert(!header.IsNull());
+    block = header;
+    block.vtx.resize(txhashes.size());
+
+    size_t tx_missing_offset = 0;
+    std::vector<PrefilledTransaction>::const_iterator prefilledit = prefilledtxn.begin();
+    size_t prefilled_count = 0, missing_count = 0, mempool_count = 0;
+    for (size_t i = 0; i < txhashes.size(); i++) {
+        if (txhashes[i].IsNull()) {
+            if (vtx_missing.size() <= tx_missing_offset)
+                return READ_STATUS_INVALID;
+            block.vtx[i] = vtx_missing[tx_missing_offset++];
+            missing_count++;
+        } else {
+            if (prefilledit != prefilledtxn.end() && prefilledit->index == i) {
+                block.vtx[i] = prefilledit->tx;
+                prefilledit++;
+                prefilled_count++;
+            } else {
+                assert(pool->lookup(txhashes[i], block.vtx[i], true));
+                mempool_count++;
+            }
+        }
+    }
+    assert(prefilledit == prefilledtxn.end());
+    if (vtx_missing.size() == tx_missing_offset) {
+        LogPrint("cmpctblock", "Successfully reconstructed block %s with %lu txn prefilled, %lu txn from mempool and %lu txn requested\n", header.GetHash().ToString(), prefilled_count, mempool_count, missing_count);
+        if (vtx_missing.size() < 5) {
+            BOOST_FOREACH(const CTransaction& tx, vtx_missing)
+                LogPrint("cmpctblock", "Reconstructed block %s required tx %s\n", header.GetHash().ToString(), tx.GetHash().ToString());
+        }
+    }
+    return vtx_missing.size() == tx_missing_offset ? READ_STATUS_OK : READ_STATUS_INVALID;
+}
+
+PartiallyDownloadedBlock::~PartiallyDownloadedBlock() {
+    if (!header.IsNull()) {
+        std::vector<PrefilledTransaction>::const_iterator prefilledit = prefilledtxn.begin();
+        for (size_t i = 0; i < txhashes.size(); i++) {
+            if (!txhashes[i].IsNull()) {
+                if (prefilledit != prefilledtxn.end() && prefilledit->index == i)
+                    prefilledit++;
+                else
+                    pool->ReleaseTxLock(txhashes[i]);
```

Comment:
Not sure what you're asking? There is no explicit lock on cs here, but ReleaseTxLock does LOCK(cs) itself.